packages/engine/Source/Scene/Model/PntsLoader.js

```diff
@@ -572,6 +582,383 @@ function makeStructuralMetadata(parsedContent, customAttributeOutput) {
   });
 }
 
+function findBoundingBox(positions) {
+  const numberOfPoints = positions.length / 3;
+  let minx = Infinity;
+  let maxx = -Infinity;
+  let miny = Infinity;
+  let maxy = -Infinity;
+  let minz = Infinity;
+  let maxz = -Infinity;
+  for (let i = 0; i < numberOfPoints; i++) {
+    if (minx > positions[3 * i]) {
+      minx = positions[3 * i];
+    }
+    if (maxx < positions[3 * i]) {
+      maxx = positions[3 * i];
+    }
+    if (miny > positions[3 * i + 1]) {
+      miny = positions[3 * i + 1];
+    }
+    if (maxy < positions[3 * i + 1]) {
+      maxy = positions[3 * i + 1];
+    }
+    if (minz > positions[3 * i + 2]) {
+      minz = positions[3 * i + 2];
+    }
+    if (maxz < positions[3 * i + 2]) {
+      maxz = positions[3 * i + 2];
+    }
+  }
+  const ret = new Float64Array(6);
+  ret[0] = minx;
+  ret[1] = maxx;
+  ret[2] = miny;
+  ret[3] = maxy;
+  ret[4] = minz;
+  ret[5] = maxz;
+  return ret;
+}
+
+function make3DTree(loader, positions) {
+  const numberOfPoints = positions.length / 3;
+  const indexedTree = new Int32Array(numberOfPoints * 2);
+  const A = new Int32Array(numberOfPoints);
+  for (let i = 0; i < 2 * numberOfPoints; i++) {
+    if (i < numberOfPoints) {
+      A[i] = i;
+    }
+    indexedTree[i] = -1;
+  }
+  const boundingBox = findBoundingBox(positions);
+  const p = partition(positions, A, 0, numberOfPoints - 1, 0);
+  loader._beginNode = A[p];
+  quicksort(positions, A, 0, p, numberOfPoints - 1, 1, indexedTree);
+  return [indexedTree, boundingBox];
+}
+
+function quicksort(positions, A, lo, mid, hi, depth, indexedTree) {
+  if (lo >= hi || lo < 0) {
+    return;
+  }
+  const p1 = partition(positions, A, lo, mid - 1, depth);
+  const p2 = partition(positions, A, mid + 1, hi, depth);
+  if (p1 > -1) {
+    indexedTree[2 * A[mid]] = A[p1];
+    quicksort(positions, A, lo, p1, mid - 1, depth + 1, indexedTree);
+  }
+  if (p2 > -1) {
+    indexedTree[2 * A[mid] + 1] = A[p2];
+    quicksort(positions, A, mid + 1, p2, hi, depth + 1, indexedTree);
+  }
+}
+
+function partition(positions, A, lo, hi, depth) {
+  const axis = depth % 3;
+  if (hi < lo) {
+    return -1;
+  }
+  if (lo === hi) {
+    return lo;
+  }
+  if (lo + 1 === hi) {
+    if (positions[3 * A[lo] + axis] > positions[3 * A[hi] + axis]) {
+      const temp = A[lo];
+      A[lo] = A[hi];
+      A[hi] = temp;
+    }
+    return hi;
+  }
+  const mid = A[Math.floor((lo + hi) / 2)];
+  const low = positions[3 * A[lo] + axis];
+  const middle = positions[3 * mid + axis];
+  const high = positions[3 * A[hi] + axis];
+
+  let pivot = A[hi];
+  let pivot_index = hi;
+  if (low < high) {
```

Comment:
nit: my personal preference would be something like:

if (low < middle && middle < high) {
} else if (low < high && high < middle) {
} else if (middle < low && low < high) {
} else if ( middle < high && high < middle) {
} else if (high < low && low < middle ) {
} else if (high < middle && middle < low) {
}

I know it's more text, but you can look at each case and know what it represents without having to unwind an unbalanced conditional treee